File: ISEmojiView/ISEmojiView.m
//
//  ISEmojiView.m
//  ISEmojiViewSample
//
//  Created by isaced on 14/12/25.
//  Copyright (c) 2014年 isaced. All rights reserved.
//

#import "ISEmojiView.h"

static const CGFloat EmojiWidth = 53;
static const CGFloat EmojiHeight = 50;
static const CGFloat EmojiFontSize = 32;

@interface ISEmojiView()<UIScrollViewDelegate>

@property (nonatomic, strong) NSArray *emojis;
@property (nonatomic, strong) UIScrollView *scrollView;
@property (nonatomic, strong) UIPageControl *pageControl;

@end

@implementation ISEmojiView

-(instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        
        // init emojis
        NSString *plistPath = [[NSBundle mainBundle] pathForResource:@"ISEmojiList" ofType:@"plist"];
        self.emojis = [NSArray arrayWithContentsOfFile:plistPath];
        
        //
        NSInteger rowNum = (CGRectGetHeight(frame) / EmojiHeight);
        NSInteger colNum = (CGRectGetWidth(frame) / EmojiWidth);
        NSInteger numOfPage = ceil((float)[self.emojis count] / (float)(rowNum * colNum));
        
        // init scrollview
        self.scrollView = [[UIScrollView alloc] initWithFrame:frame];
        self.scrollView.pagingEnabled = YES;
        self.scrollView.showsHorizontalScrollIndicator = NO;
        self.scrollView.showsVerticalScrollIndicator = NO;
        self.scrollView.delegate = self;
        self.scrollView.contentSize = CGSizeMake(CGRectGetWidth(frame) * numOfPage,
                                                 CGRectGetHeight(frame));
        [self addSubview:self.scrollView];
        
        // add emojis
        
        NSInteger row = 0;
        NSInteger column = 0;
        NSInteger page = 0;
        
        NSInteger emojiPointer = 0;
        for (int i = 0; i < [self.emojis count] + numOfPage - 1; i++) {
            
            // Pagination
            if (i % (rowNum * colNum) == 0) {
                page ++;    // Increase the number of pages
                row = 0;    // the number of lines is 0
                column = 0; // the number of columns is 0
            }else if (i % colNum == 0) {
                // NewLine
                row += 1;   // Increase the number of lines
                column = 0; // The number of columns is 0
            }
            
            CGRect currentRect = CGRectMake(((page-1) * frame.size.width) + (column * EmojiWidth),
                                            row * EmojiHeight,
                                            EmojiWidth,
                                            EmojiHeight);
            
            if (row == (rowNum - 1) && column == (colNum - 1)) {
                // last position of page, add delete button
                
                ISDeleteButton *deleteButton = [ISDeleteButton buttonWithType:UIButtonTypeCustom];
                [deleteButton addTarget:self action:@selector(deleteButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
                deleteButton.frame = currentRect;
                deleteButton.tintColor = [UIColor blackColor];
                [self.scrollView addSubview:deleteButton];
                
            }else{
                NSString *emoji = self.emojis[emojiPointer++];
                
                // init Emoji Button
                UIButton *emojiButton = [UIButton buttonWithType:UIButtonTypeCustom];
                emojiButton.titleLabel.font = [UIFont fontWithName:@"Apple color emoji" size:EmojiFontSize];
                [emojiButton setTitle:emoji forState:UIControlStateNormal];
                [emojiButton addTarget:self action:@selector(emojiButtonPressed:) forControlEvents:UIControlEventTouchUpInside];
                
                emojiButton.frame = currentRect;
                [self.scrollView addSubview:emojiButton];
            }
            
            column++;
        }
        
        // add PageControl
        self.pageControl = [[UIPageControl alloc] init];
        self.pageControl.hidesForSinglePage = YES;
        self.pageControl.currentPage = 0;
        self.pageControl.backgroundColor = [UIColor clearColor];
        self.pageControl.numberOfPages = numOfPage;
        CGSize pageControlSize = [self.pageControl sizeForNumberOfPages:numOfPage];
        self.pageControl.frame = CGRectMake(CGRectGetMidX(frame) - (pageControlSize.width / 2),
                                            CGRectGetHeight(frame) - pageControlSize.height + 5,
                                            pageControlSize.width,
                                            pageControlSize.height);
        [self.pageControl addTarget:self action:@selector(pageControlTouched:) forControlEvents:UIControlEventValueChanged];
        [self addSubview:self.pageControl];
        

    }
    return self;
}

- (void)pageControlTouched:(UIPageControl *)sender {
    CGRect bounds = self.scrollView.bounds;
    bounds.origin.x = CGRectGetWidth(bounds) * sender.currentPage;
    [self.scrollView scrollRectToVisible:bounds animated:YES];
}

- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    CGFloat pageWidth = CGRectGetWidth(scrollView.frame);
    NSInteger newPageNumber = floor((scrollView.contentOffset.x - pageWidth / 2) / pageWidth) + 1;
    if (self.pageControl.currentPage == newPageNumber) {
        return;
    }
    self.pageControl.currentPage = newPageNumber;
}

- (void)emojiButtonPressed:(UIButton *)button {
    
    // Add a simple scale animation
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"transform.scale"];
    animation.byValue = @0.3;
    animation.duration = 0.1;
    animation.autoreverses = YES;
    [button.layer addAnimation:animation forKey:nil];
    
    // Callback
    if ([self.delegate respondsToSelector:@selector(emojiView:didSelectEmoji:)]) {
        [self.delegate emojiView:self didSelectEmoji:button.titleLabel.text];
    }
}

- (void)deleteButtonPressed:(UIButton *)button{
    // Add a simple scale animation
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"transform.scale"];
    animation.toValue = @0.9;
    animation.duration = 0.1;
    animation.autoreverses = YES;
    [button.layer addAnimation:animation forKey:nil];
    
    // Callback
    if ([self.delegate respondsToSelector:@selector(emojiView:didPressDeleteButton:)]) {
        [self.delegate emojiView:self didPressDeleteButton:button];
    }
}

@end

@implementation ISDeleteButton

-(void)drawRect:(CGRect)rect{

    // Rectangle Drawing
    UIBezierPath* rectanglePath = UIBezierPath.bezierPath;
    [rectanglePath moveToPoint: CGPointMake(5, 25.05)];
    [rectanglePath addLineToPoint: CGPointMake(20.16, 36)];
    [rectanglePath addLineToPoint: CGPointMake(45.5, 36)];
    [rectanglePath addLineToPoint: CGPointMake(45.5, 13.5)];
    [rectanglePath addLineToPoint: CGPointMake(20.16, 13.5)];
    [rectanglePath addLineToPoint: CGPointMake(5, 25.05)];
    [rectanglePath closePath];
    [self.tintColor setStroke];
    rectanglePath.lineWidth = 1;
    [rectanglePath stroke];
    
    
    // Bezier Drawing
    UIBezierPath* bezierPath = UIBezierPath.bezierPath;
    [bezierPath moveToPoint: CGPointMake(26.5, 20)];
    [bezierPath addLineToPoint: CGPointMake(36.5, 29.5)];
    [bezierPath moveToPoint: CGPointMake(36.5, 20)];
    [bezierPath addLineToPoint: CGPointMake(26.5, 29.5)];
    [self.tintColor setStroke];
    bezierPath.lineWidth = 1;
    [bezierPath stroke];
}

@end

File: ISEmojiView/Classes/ISEmojiView.swift
//
//  ISEmojiView.swift
//  Pods
//
//  Created by isaced on 2016/11/8.
//
//

import Foundation

/// emoji view action callback delegate
public protocol ISEmojiViewDelegate: class {
    
    /// did press a emoji button
    ///
    /// - Parameters:
    ///   - emojiView: the emoji view
    ///   - emoji: a emoji
    func emojiViewDidSelectEmoji(emojiView: ISEmojiView, emoji: String)
    
    
    /// will delete last character in you input view
    ///
    /// - Parameter emojiView: the emoji view
    func emojiViewDidPressDeleteButton(emojiView: ISEmojiView)
}

fileprivate let EmojiSize = CGSize(width: 38, height: 38)
fileprivate let EmojiFontSize = CGFloat(35.0)
fileprivate let collectionInset = UIEdgeInsets(top: 10, left: 10, bottom: 35, right: 10)
fileprivate let collectionMinimumLineSpacing = CGFloat(0)
fileprivate let collectionMinimumInteritemSpacing = CGFloat(8.0)

/// A emoji keyboard view
public class ISEmojiView: UIView, UICollectionViewDataSource, UICollectionViewDelegate {

    var defaultFrame: CGRect {
        return CGRect(x: 0, y: 0, width: UIScreen.main.bounds.width, height: 216)
    }
    var collectionView: UICollectionView = {
        let layout = UICollectionViewFlowLayout()
        layout.itemSize = EmojiSize
        layout.scrollDirection = .horizontal
        layout.minimumLineSpacing = collectionMinimumLineSpacing
        layout.minimumInteritemSpacing = collectionMinimumInteritemSpacing
        layout.sectionInset = collectionInset
        let collection = UICollectionView(frame: .zero, collectionViewLayout: layout)
        collection.showsHorizontalScrollIndicator = false
        collection.showsVerticalScrollIndicator = false
        collection.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        collection.backgroundColor = UIColor(red: 249/255.0, green: 249/255.0, blue: 249/255.0, alpha: 1)
        collection.register(ISEmojiCell.self, forCellWithReuseIdentifier: "cell")
        return collection
    }()
    var pageControl: UIPageControl = {
        let pageContr = UIPageControl()
        pageContr.hidesForSinglePage = true
        pageContr.currentPage = 0
        pageContr.backgroundColor = .clear
        return pageContr
    }()
    var deleteButton: UIButton = {
        let button = UIButton(type: .system)
        let deleteButtonImage = UIImage(named: "icon_delete", in: ISEmojiView.thisBundle(), compatibleWith: nil)
        button.setImage(deleteButtonImage, for: .normal)
        button.tintColor = .lightGray
        return button
    }()
    var emojis: [[String]] = {
        if let filePath = ISEmojiView.pathOfResourceInBundle(filename: "ISEmojiList", filetype: "plist") {
            if let sections = NSDictionary(contentsOfFile: filePath) as? [String:[String]] {
                var emojiList: [[String]] = []
                for sectionName in ["People","Nature","Objects","Places","Symbols"] {
                    emojiList.append(sections[sectionName]!)
                }
                return emojiList
            }
        }
        return []
    }()
    public weak var delegate: ISEmojiViewDelegate?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
    }
    
    required public init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    public override func layoutSubviews() {
        updateControlLayout()
    }

    func setupUI() {
        frame = defaultFrame
        
        // ScrollView
        collectionView.frame = self.bounds
        collectionView.dataSource = self
        collectionView.delegate = self
        addSubview(collectionView)
        collectionView.reloadData()
        
        // PageControl
        pageControl.addTarget(self, action: #selector(pageControlTouched), for: .touchUpInside)
        pageControl.pageIndicatorTintColor = UIColor(red: 229/255.0, green: 229/255.0, blue: 229/255.0, alpha: 1)
        pageControl.currentPageIndicatorTintColor = UIColor(red: 153/255.0, green: 153/255.0, blue: 153/255.0, alpha: 1)
        pageControl.autoresizingMask = [.flexibleTopMargin, .flexibleLeftMargin, .flexibleRightMargin]
        addSubview(pageControl)
        
        // DeleteButton
        deleteButton.addTarget(self, action: #selector(deleteButtonPressed), for: .touchUpInside)
        addSubview(deleteButton)
    }
    
    func updateControlLayout() {
        frame = defaultFrame
        
        // update page control
        let pageCount = Int(ceil(collectionView.contentSize.width / collectionView.bounds.width))
        let pageControlSizes = pageControl.size(forNumberOfPages: pageCount)
        pageControl.frame = CGRect(x: frame.midX - pageControlSizes.width / 2.0,
                                        y: frame.height-pageControlSizes.height,
                                        width: pageControlSizes.width,
                                        height: pageControlSizes.height)
        pageControl.numberOfPages = pageCount
        
        // update delete button
        deleteButton.frame = CGRect(x: frame.width - 45, y: frame.height - 40, width: 40, height: 40)
    }
    
    //MARK: <UICollectionView>
    public func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 5
    }
    
    public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 50
    }
    
    public func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "cell", for: indexPath) as! ISEmojiCell
        cell.setEmoji(emojis[indexPath.section][indexPath.row])
        return cell
    }
    
    public func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let emoji = emojis[indexPath.section][indexPath.row]
        self.delegate?.emojiViewDidSelectEmoji(emojiView: self, emoji: emoji)
    }
    
    //MARK: <UIScrollView>
    
    public func scrollViewDidScroll(_ scrollView: UIScrollView) {
        let pageWidth = scrollView.bounds.width
        let newPageNumber = Int(floor((scrollView.contentOffset.x - pageWidth / 2.0) / pageWidth) + 1)
        if pageControl.currentPage != newPageNumber {
            pageControl.currentPage = newPageNumber
        }
    }
    
    //MARK: Action
    
    func emojiButtonPressed(sender: UIButton) {
        if let emoji = sender.titleLabel?.text {
            self.delegate?.emojiViewDidSelectEmoji(emojiView: self, emoji: emoji)
        }
    }
    
    func deleteButtonPressed(sender: UIButton) {
        self.delegate?.emojiViewDidPressDeleteButton(emojiView: self)
    }
    
    func pageControlTouched(sender: UIPageControl) {
        var bounds = collectionView.bounds
        bounds.origin.x = bounds.width * CGFloat(sender.currentPage)
        collectionView.scrollRectToVisible(bounds, animated: true)
    }
    
    //MARK: Tools
    
    static func thisBundle() -> Bundle {
        let podBundle = Bundle(for: ISEmojiView.classForCoder())
        if let bundleURL = podBundle.url(forResource: "ISEmojiView", withExtension: "bundle") {
            if let bundle = Bundle(url: bundleURL) {
                return bundle
            }else{
                assertionFailure("Could not load the bundle")
            }
        }else{
            assertionFailure("Could not create a path to the bundle")
        }
        return Bundle()
    }
    static func pathOfResourceInBundle(filename:String, filetype: String) -> String? {
        if let filePath = thisBundle().path(forResource: filename, ofType: filetype) {
            return filePath
        }
        return nil
    }
}

/// the emoji cell in the grid
fileprivate class ISEmojiCell: UICollectionViewCell {
    
    var emojiLabel: UILabel = {
        let label = UILabel()
        label.font = UIFont(name: "Apple color e  moji", size: EmojiFontSize)
        label.textAlignment = .center
        label.autoresizingMask = [.flexibleWidth,.flexibleHeight]
        return label
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setupUI()
    }
    
    func setupUI() {
        emojiLabel.frame = bounds
        self.addSubview(emojiLabel)
    }
    
    func setEmoji(_ emoji: String) {
        self.emojiLabel.text = emoji
    }
}
